src/cryptonote_core/blockchain.cpp

```diff
@@ -1779,14 +1779,13 @@ bool Blockchain::handle_alternative_block(const block& b, const crypto::hash& id
             break;
           }
         }
-        get_altblock_longhash(bei.bl, proof_of_work, get_current_blockchain_height(), bei.height, seedheight, seedhash);
       } else
       {
-        goto mainhash;
+        seedhash = get_block_id_by_height(seedheight);
       }
+      get_altblock_longhash(bei.bl, proof_of_work, get_current_blockchain_height(), bei.height, seedheight, seedhash);
```

Comment:
Why does this use the blockchain height ? It should never need to.

Reply:
That's used to identify which is the current cache, so that the alt hasher can use the non-current cache.

Reply:
Ah, that makes sense, thanks.

Reply:
Is it always guaranteed that alt_chain's height is larger than current main chain height? Because if it's not, we can end up with uninitialized seed_hash here.

Reply:
In general, it will not be.

Reply:
Alt chain can both start and end before current seedheight, so this "for" loop is not guaranteed to initialize seedhash properly. An example of such altchain is "frozen timestamp" attack alt chain where blocks are very slow but cumulative difficulty overtakes current main chain.

Reply:
@hyc This "for" loop should start from the latest common seedhash, not from alt chain begin, and go up to alt chain's end to find correct seedhash.

Reply:
Adding something like
```
if (seed_hash == null_hash)
    seed_hash = get_block_id_by_height(std::min(seedheight, rx_seedheight(alt_chain.front().height)));
```
should be enough.